src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1357,156 @@ internal bool CanAddValue(HttpHeaderParser parser)
         }
         #endregion
     }
+
+    public readonly struct HttpHeadersNonValidated : IEnumerable<KeyValuePair<string, HeaderStringValues>>
+    {
+        private readonly IEnumerable<KeyValuePair<HeaderDescriptor, object>>? _headerStore;
+
+        internal HttpHeadersNonValidated(IEnumerable<KeyValuePair<HeaderDescriptor, object>>? headerStore)
+        {
+            _headerStore = headerStore;
+        }
+
+        public IEnumerator<KeyValuePair<string, HeaderStringValues>> GetEnumerator() => _headerStore != null
```

Comment:
Couldn't the `Enumerator` itself handle the `_headerStore` being `null` instead of the `Array.Empty.GetEnumerator()` here? Our enumerator is a struct so it might even be cheaper.

Reply:
This mirrors the current implementation in HttpHeaders.GetEnumerator.

Reply:
> This mirrors the current implementation in HttpHeaders.GetEnumerator.

But the HttpHeaders API returns the interface.  This needs to be strongly-typed as returning the struct.  That changes things.

Reply:
Done.